(fix) CodeActAgent/LLM: react on should_exit flag (user cancellation)

tobitege opened the pull request:
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

CodeActAgent and LLM completion shall react on should_exit flag (user cancellation)

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

Retry process on llm completion attempt now checks the recently introduced `should_exit` flag and raises an `OperationCancelled` to end retries.
- LLM: replaced `after` with `before_sleep` retry param (only triggered on fail now)
- LLM: added `_call_completion` to encapsulate `litellm_completion` and is easier to mock in tests
- LLM: changed logging of token counts to be on a single line, like cost ist
- use newly added `OperationCancelled` exception (interim was `UserCancelledError`) to exit retry loop
- extended `test_llm.py` tests with several exception handling tests

Fixes #3965 

Added tests:
```
tests/unit/test_llm.py::test_completion PASSED
tests/unit/test_llm.py::test_completion_retries[APIConnectionError-extra_args0-2] PASSED
tests/unit/test_llm.py::test_completion_retries[ContentPolicyViolationError-extra_args1-2] PASSED
tests/unit/test_llm.py::test_completion_retries[InternalServerError-extra_args2-2] PASSED
tests/unit/test_llm.py::test_completion_retries[OpenAIError-extra_args3-2] PASSED
tests/unit/test_llm.py::test_completion_retries[RateLimitError-extra_args4-2] PASSED
tests/unit/test_llm.py::test_completion_rate_limit_wait_time PASSED
tests/unit/test_llm.py::test_completion_exhausts_retries PASSED
tests/unit/test_llm.py::test_completion_operation_cancelled PASSED
tests/unit/test_llm.py::test_completion_keyboard_interrupt PASSED
tests/unit/test_llm.py::test_completion_keyboard_interrupt_handler PASSED
```

enyst commented on openhands/llm/llm.py:
```diff
@@ -207,14 +211,21 @@ def custom_completion_wait(retry_state):
-            retry=retry_if_exception_type(self.retry_exceptions),
+            retry=(
+                retry_if_exception_type(self.retry_exceptions)
+                & retry_if_not_exception_type(UserCancelledError)
```
just a bit curious here, `self.retry_exceptions` is [this](https://github.com/All-Hands-AI/OpenHands/pull/3968/files#diff-b46ec9dd4b319d75809520e34b2559fd2ce46ca4895709eae9312aa04a8fa9a4R89) if I see this right? it doesn't have UserCancelledError - of course we don't want to. So the second part will never be hit

tobitege replied:
The 2nd part will be hit from the `before_sleep` which does end the retry loop.

enyst commented on agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -210,6 +211,8 @@ def step(self, state: State) -> Action:
+            raise e
         except Exception as e:
             logger.error(f'{e}')
             error_message = '{}: {}'.format(type(e).__name__, str(e).split('\n')[0])
```
Oh, not in scope of this PR, but this doesn't look quite right. IMHO we shouldn't have needed to add this kind of exception in the agent, I don't think, it should have simply fallen down to the controller. But it won't because we catch everything and not re-raise.
One more consequence of this code here, is that now _every_ agent needs to handle ctrl+c, and that cannot be right...

Perhaps we can add a TODO: fix exception handling?

tobitege replied:
Yeh, we need a separate issue for that.

enyst commented on openhands/llm/llm.py:
```diff
@@ -179,6 +181,8 @@ def attempt_on_error(retry_state):
         def custom_completion_wait(retry_state):
             """Custom wait function for litellm completion."""
+            if should_exit():
+                raise UserCancelledError('Shutdown requested during retry by user.')
```
Could you please help me understand why does this happen now, what breaks it. Why do we need to add a custom mechanism if we didn't before?

tobitege replied:
This has been moved to the `before_sleep` handler, which is called _before_ `custom_completion_wait` to end the loop.

enyst commented:
Thanks for looking into this! It's late and I may not get to test properly, but at first sight it seems to work. I don't see the underlying reason why this fix seems necessary, though.

tobitege commented:
- LLM: replaced `after` with `before_sleep` retry param
- use new `OperationCancelled` exception instead of `UserCancelledError`
- extended `test_llm.py` tests with exception handling tests
- changed logging of token counts to be on a single line, like cost

@enyst - Thanks for the valuable feedback, I'll try to address it here in one comment instead of spread out:
The long-time use of the `after` in the decorator was not quite optimal, as it is called for BOTH successes and fails.

The circumstance, that the keyboard interrupt (like logged in your PR) didn't "make it through" is, imo,
a side effect of the new shutdown handlers, not the LLM's `@retry` decorator.
E.g. `KeyboardInterrupt` wouldn't have been intercepted either way as it is not in the list of retried exceptions.

The remedy here was to use the `should_exit()` method and then to raise a (new) `OperationCancelled`
exception, that ends the retry loop (`UserCancelledError` was inappropriate, as it could be used in relation to agents).

I've added multiple new tests to 'test_llm.py' to test the exception handling, incl. keyboard interrupt (all successful in my WSL).

The exception handling in CodeActAgent was required - and probably needs to be moved to agent_controller? - for
the fact that if an exception got through, it broke everything.

enyst commented on openhands/core/exceptions.py:
```diff
@@ -77,3 +77,10 @@ def __init__(self, message='User cancelled the request'):
+class OperationCancelled(Exception):
+    """Exception raised when an operation is cancelled (e.g. by a keyboard interrupt)."""
+
+    def __init__(self, message='Operation was cancelled'):
```
Just a question here, why aren't we still using UserCancelled, as previously?

tobitege replied:
It's aimed to be future-proof: if `UserCancelled` were to be used e.g. for a UI button to cancel an operation, it is not the same in this context.

enyst replied:
If I see this right, UserCancelledError is not used, but in a test it is assumed to be what you describe.

tobitege replied:
Ah, that's the async variant, which we don't use yet anyways. I'll change those occurences as well.

enyst replied:
I don't think we need to change them? Wait, so in my understanding, it was assumed in that test that it means "a UI button to cancel an operation", the kind of Stop Task coming from somewhere that doesn't shut down the app. Then it makes sense that we keep UserCancelledError _for that_. I thought that's what you said too.

tobitege replied:
Yeh, I was just looking at the callback function, so those UserCancelledError occurences are actually desired, so no changes then. :)

enyst replied:
On a side note, I really think we'd be happier if we move the async and streaming code somehow, in a subclass or a mixin or something. I know they will be needed, but they currently aren't in use, and I don't know, it's just difficult to read llm.py 😢

tobitege replied:
But not in this PR, right?

enyst commented on openhands/llm/llm.py:
```diff
@@ -169,13 +174,18 @@ def __init__(
-            """Custom attempt function for litellm completion."""
+        def log_retry_attempt(retry_state):
+            """With before_sleep, this is called before `custom_completion_wait` and
+            ONLY if the retry is triggered by an exception."""
```
"ONLY if the retry is triggered by an exception" - isn't the retry always triggered by an exception?

tobitege replied:
The `before_sleep` only fires on exceptions, the former `after` fired for every attempt, even the succesful one.

enyst commented on openhands/llm/llm.py:
```diff
@@ -447,7 +466,16 @@ def _format_content_element(self, element):
+        makes it mockable for testing.
+        """
+        return litellm.completion(*args, **kwargs)
+
```
I feel like it's possible we need this, but I'm wondering, we have a mock for `completion` in tests. Isn't it possible that we use that instead of wrapping one more time?

It's absolutely great that this PR adds tests for this retry. It helps, it can simplify our digging in here immensely.

tobitege replied:
If you mock LLM's `completion`, you can't test its inside code. With mocking this encapsulation, you have way more options for testing.

enyst replied:
But the retries mechanism is a wrapper itself, around completion? It's not inside code.

Please don't get me wrong, I think adding retries to tests is something we should do, it's just too good - and necessary. I just hope that we can test retries without another wrapper in llm.py. Maybe I just don't see clearly enough why it's needed?

tobitege replied:
> But the retries mechanism is a wrapper itself, around completion? It's not inside code.
> 
> Please don't get me wrong, I think adding retries to tests is something we should do, it's just too good - and necessary. I just hope that we can test retries without another wrapper in llm.py. Maybe I just don't see clearly enough why it's needed?

Working on it now... :)

tobitege replied:
@enyst - Done! Reverted that extra wrapper method and changed the tests accordingly.
Imo, this should be good for final review now. 🙏

enyst commented:
@tobitege Just to make sure I understand correctly: the retries in llm.py are a regression introduced by https://github.com/All-Hands-AI/OpenHands/pull/3926, or were they broken before, and 3926 didn't fix it since it added `should_continue` only to the async variant?

We can see how 3926 implemented its new calls, it's without throwing and catching a special exception, can we perhaps do that here? 🤔 I would love that, at least for now, because it would mean that we don't need to do stuff in the agents' code which, as we talked there, shouldn't be necessary in the first place anyway.

tobitege commented:
> @tobitege Just to make sure I understand correctly: the retries in llm.py are a regression introduced by #3926, or were they broken before, and 3926 didn't fix it since it added `should_continue` only to the async variant? Cc: @tofarr what do you think?
> 
> We can see how 3926 implemented its new calls, it's without throwing and catching a special exception, can we perhaps do that here? 🤔 I would love that, at least for now, because it would mean that we don't need to do stuff in the agents' code which, as we talked there, shouldn't be necessary in the first place anyway.

I think that the CodeActAgent exception handler was the main issue as it previously caught too many exceptions in order to not break the whole agent.
Now, with the applied changes, using the shutdown handler method, any related exception is raised.
